src/main/scala/org/constellation/snapshotstreaming/storage/FileBasedLastIncrementalGlobalSnapshotStorage.scala

```diff
@@ -0,0 +1,50 @@
+package org.constellation.snapshotstreaming.storage
+
+import cats.effect.Async
+import cats.effect.std.Semaphore
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.merkletree.StateProofValidator
+import org.tessellation.node.shared.domain.snapshot.storage.LastSnapshotStorage
+import org.tessellation.schema._
+import org.tessellation.security._
+import fs2.io.file._
+import io.circe.Codec
+import io.circe.generic.semiauto.deriveCodec
+
+class FileBasedLastIncrementalGlobalSnapshotStorage[F[_]: Async: HasherSelector](semaphore: Semaphore[F], path: Path)(
+  implicit codec: Codec[SnapshotWithState[GlobalIncrementalSnapshot, GlobalSnapshotInfo]]
+) extends BaseLastIncrementalSnapshotStorage[
+      F,
+      GlobalSnapshotStateProof,
+      GlobalIncrementalSnapshot,
+      GlobalSnapshotInfo
+    ](semaphore, path) {
+
+  protected def validateStateProof(snapshot: Hashed[GlobalIncrementalSnapshot], state: GlobalSnapshotInfo): F[Unit] =
+    HasherSelector[F].forOrdinal(snapshot.ordinal) { implicit hasher =>
+      (hasher.getLogic(snapshot.ordinal) match {
+        case JsonHash => StateProofValidator.validate(snapshot, state)
+        case KryoHash => StateProofValidator.validate(snapshot, GlobalSnapshotInfoV2.fromGlobalSnapshotInfo(state))
+      })
+        .map(_.isValid)
+        .flatMap(new Throwable("State proof doesn't match!").raiseError[F, Unit].unlessA)
+    }
+
+}
+
+object FileBasedLastIncrementalGlobalSnapshotStorage {
+
+  def make[F[_]: Async: Files: KryoSerializer: HasherSelector](
+    path: Path
+  ): F[LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo]] =
+    Semaphore[F](1).map { semaphore =>
```

Comment:
Same comment here as earlier WRT using a mutex; in this case I can see the storage is passed to another service.